File: javaspider/src/main/java/com/javaspider/performance/circuit/CircuitBreaker.java
package com.javaspider.performance.circuit;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * 熔断器实现
 * 三种状态：CLOSED（正常）、OPEN（熔断）、HALF_OPEN（半开）
 */
public class CircuitBreaker {
    
    public enum State {
        CLOSED,     // 正常状态，允许请求通过
        OPEN,       // 熔断状态，拒绝请求
        HALF_OPEN   // 半开状态，允许部分请求通过测试
    }
    
    private final int failureThreshold;      // 失败阈值
    private final int successThreshold;      // 成功阈值（半开状态）
    private final long timeoutMs;            // 超时时间（毫秒）
    private final AtomicInteger failureCount;
    private final AtomicInteger successCount;
    private final AtomicLong lastFailureTime;
    private volatile State state;
    
    public CircuitBreaker(int failureThreshold, int successThreshold, long timeoutMs) {
        this.failureThreshold = failureThreshold;
        this.successThreshold = successThreshold;
        this.timeoutMs = timeoutMs;
        this.failureCount = new AtomicInteger(0);
        this.successCount = new AtomicInteger(0);
        this.lastFailureTime = new AtomicLong(0);
        this.state = State.CLOSED;
    }
    
    public CircuitBreaker() {
        this(5, 3, 60000); // 默认：5 次失败，3 次成功，60 秒超时
    }
    
    /**
     * 检查是否允许请求通过
     */
    public synchronized boolean allowRequest() {
        long now = System.currentTimeMillis();
        
        if (state == State.OPEN) {
            // 检查是否超过超时时间
            if (now - lastFailureTime.get() > timeoutMs) {
                state = State.HALF_OPEN;
                successCount.set(0);
                return true;
            }
            return false;
        }
        
        return true;
    }
    
    /**
     * 记录成功请求
     */
    public synchronized void recordSuccess() {
        if (state == State.HALF_OPEN) {
            int count = successCount.incrementAndGet();
            if (count >= successThreshold) {
                state = State.CLOSED;
                failureCount.set(0);
                successCount.set(0);
            }
        } else if (state == State.CLOSED) {
            failureCount.set(0);
        }
    }
    
    /**
     * 记录失败请求
     */
    public synchronized void recordFailure() {
        long now = System.currentTimeMillis();
        lastFailureTime.set(now);
        
        if (state == State.HALF_OPEN) {
            state = State.OPEN;
            successCount.set(0);
        } else if (state == State.CLOSED) {
            int count = failureCount.incrementAndGet();
            if (count >= failureThreshold) {
                state = State.OPEN;
            }
        }
    }
    
    /**
     * 获取当前状态
     */
    public State getState() {
        return state;
    }
    
    /**
     * 重置熔断器
     */
    public synchronized void reset() {
        state = State.CLOSED;
        failureCount.set(0);
        successCount.set(0);
        lastFailureTime.set(0);
    }
    
    /**
     * 获取统计信息
     */
    public String getStats() {
        return String.format("CircuitBreaker{state=%s, failures=%d, successes=%d}",
                state, failureCount.get(), successCount.get());
    }
}

File: pyspider/performance/circuit_breaker.py
"""
PySpider 熔断器实现
"""

import time
import threading
from enum import Enum
from typing import Optional


class CircuitState(Enum):
    CLOSED = "closed"  # 正常状态
    OPEN = "open"  # 熔断状态
    HALF_OPEN = "half_open"  # 半开状态


class CircuitBreaker:
    """
    熔断器实现

    使用示例:
        breaker = CircuitBreaker(failure_threshold=5, success_threshold=3, timeout=60)

        def make_request():
            if not breaker.allow_request():
                raise Exception("Circuit is open")
            try:
                result = do_something()
                breaker.record_success()
                return result
            except Exception as e:
                breaker.record_failure()
                raise
    """

    def __init__(
        self,
        failure_threshold: int = 5,
        success_threshold: int = 3,
        timeout: float = 60.0,
        name: Optional[str] = None,
    ):
        self.failure_threshold = failure_threshold
        self.success_threshold = success_threshold
        self.timeout = timeout
        self.name = name or "circuit_breaker"

        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time: Optional[float] = None
        self._lock = threading.Lock()

    @property
    def state(self) -> CircuitState:
        with self._lock:
            if self._state == CircuitState.OPEN and self._last_failure_time is not None:
                if time.time() - self._last_failure_time > self.timeout:
                    self._state = CircuitState.HALF_OPEN
                    self._success_count = 0
        return self._state

    @property
    def failure_count(self) -> int:
        return self._failure_count

    @property
    def success_count(self) -> int:
        return self._success_count

    def allow_request(self) -> bool:
        """检查是否允许请求通过"""
        with self._lock:
            if self._state == CircuitState.OPEN:
                # 检查是否超过超时时间
                if self._last_failure_time is None:
                    return True

                elapsed = time.time() - self._last_failure_time
                if elapsed > self.timeout:
                    # 转换为半开状态
                    self._state = CircuitState.HALF_OPEN
                    self._success_count = 0
                    return True
                return False

            return True

    def record_success(self):
        """记录成功请求"""
        with self._lock:
            if self._state == CircuitState.OPEN and self._last_failure_time is not None:
                if time.time() - self._last_failure_time > self.timeout:
                    self._state = CircuitState.HALF_OPEN
                    self._success_count = 0
            if self._state == CircuitState.HALF_OPEN:
                self._success_count += 1
                if self._success_count >= self.success_threshold:
                    # 恢复到关闭状态
                    self._state = CircuitState.CLOSED
                    self._failure_count = 0
                    self._success_count = 0
            elif self._state == CircuitState.CLOSED:
                self._failure_count = 0

    def record_failure(self):
        """记录失败请求"""
        with self._lock:
            self._last_failure_time = time.time()

            if self._state == CircuitState.HALF_OPEN:
                # 回到熔断状态
                self._state = CircuitState.OPEN
                self._success_count = 0
            elif self._state == CircuitState.CLOSED:
                self._failure_count += 1
                if self._failure_count >= self.failure_threshold:
                    # 进入熔断状态
                    self._state = CircuitState.OPEN

    def reset(self):
        """重置熔断器"""
        with self._lock:
            self._state = CircuitState.CLOSED
            self._failure_count = 0
            self._success_count = 0
            self._last_failure_time = None

    def get_stats(self) -> dict:
        """获取统计信息"""
        return {
            "name": self.name,
            "state": self._state.value,
            "failure_count": self._failure_count,
            "success_count": self._success_count,
            "last_failure_time": self._last_failure_time,
        }

    def __repr__(self) -> str:
        return (
            f"CircuitBreaker(state={self._state.value}, failures={self._failure_count})"
        )


class CircuitBreakerOpen(Exception):
    """熔断器打开异常"""

    pass
